aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -190,6 +215,46 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    fun ed25519_verify_t<T: drop> (signature: vector<u8>, public_key: vector<u8>, data: T): bool {
+        let encoded = Proof {
+            type_info: type_info::type_of<T>(),
+            inner: data,
+        };
+        signature::ed25519_verify(signature, public_key, bcs::to_bytes(&encoded))
+    }
+
+    public entry fun rotate_authentication_key_ed25519(account: &signer, new_public_key: vector<u8>, signature: vector<u8>) acquires Account, OriginatingAddress {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr), error::not_found(EACCOUNT));
+        assert!(
+            vector::length(&new_public_key) == 32,
+            error::invalid_argument(EMALFORMED_PUBLIC_KEY)
+        );
+        assert!(
+            vector::length(&signature) == 64,
+            error::invalid_argument(EMALFORMED_PROOF_OF_KNOWLEDGE)
+        );
+
+        let account_resource = borrow_global_mut<Account>(addr);
+        let current_auth_key = create_address(account_resource.authentication_key);
+
+        let proof= RotationProof {
+            originator: addr,
+            current_auth_key: current_auth_key,
+        };
+
+        assert!(ed25519_verify_t(signature, new_public_key, proof), EINVALID_PROOF_OF_KNOWLEDGE);
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        if (table::contains(address_map, current_auth_key)) {
+            table::remove(address_map, current_auth_key);
```

Comment:
This is fine outside of smart contracts but can cost extra gas unnecessarily here. This would be cheaper: If it exists, modifying it directly. Otherwise, add it.

Reply:
makes sense! will update this : )

Reply:
oh wait, in this case we're changing the key so we actually need to remove it and add a new key pair back?